lips/lip-4.md

```diff
@@ -0,0 +1,222 @@
+
+
+---
+# Abstract / Motivation
+---
+
+Merchants and wallet holders may rely on third parties to maintain custody of their wallet and may optionally want to receive additional data in transactions.  Libra leverages subaddressing to provide this functionality.  This document describes an approach to supporting transactions like this by defining a standard that supports:
+
+* Subaddressing to support multiplexing a single account address into multiple wallets, enabling custodial wallets to maintain a single address rather than maintain numerous addresses.
+
+As an example, assume a recipient account provides custodial wallets. To send funds to this wallet within Libra:
+
+* The recipient crafts a QR code containing a URI that references the account hosting the custodial wallet, or the host, and a reference to their wallet.
+* The consumer scans the QR code and constructs a transaction and includes within it the subaddress.
+* Once the transaction has been submitted and committed to the blockchain, the host notifies the recipient of the transaction.
+
+---
+# Specification
+---
+
+# The Lifetime of a Transaction Containing Metadata
+
+The first step to submitting a transaction is producing the metadata. The sender first produces a *Libra Canonically Serialized (LCS)* metadata_wrapper consisting of an LCS-serialized MetadataType object:
+
+```
+enum MetadataType {
+  Undefined,
+  GeneralMetadataType(GeneralMetadata),
+  TravelRuleMetadataType(TravelRuleMetadata),
+  UnstructuredStringMetadataType(UnstructuredStringMetadata),
+}
+
+// Used for versioning of general metadata
+enum GeneralMetadata {
+    GeneralMetadataVersion0(GeneralMetadatav0),
+}
+
+struct GeneralMetadatav0 {
+    // Subaddress to which the funds are being sent
+    Option<Vec<u8>> to_subaddress,
+    
+    // Subaddress from which the funds are being sent
+    Option<Vec<u8>> from_subaddress,
+    
+    // Event sequence number of referenced payment
+    Option<u64> referenced_event,
+}
+
+// Used for versioning of travel rule metadata
+enum TravelRuleMetadata {
+    TravelRuleMetadataVersion0(TravelRuleMetadatav0),
+}
+
+struct TravelRuleMetadatav0 {
+    // Off-chain reference_id.  Used when off-chain APIs are used.
+    // Specifies the off-chain reference ID that was agreed upon in off-chain APIs.
+    Option<String> off_chain_reference_id,
+}
+
+struct UnstructuredStringMetadata {
+    // Unstructured string metadata
+    Option<String> metadata,
+}
+```
+
+
+Using the initial example described in the motivation, the merchant whose wallet, 0x5678, is hosted by 0x1234 may post a URI for shoes that cost 20 microlibra.  The purchaser would then submit a transaction containing the following metadata:
```

Comment:
- I didn't understand what 0x1234 was and why it didn't appear in the metadata
- 0x5678 got a screamin' deal on those shoes